packages/app/src-tauri/src/commands/transcription.rs

```diff
@@ -0,0 +1,212 @@
+        let canc_token: CancellationToken = receiver.into();
+
+        if let Err(err) = model.transcribe_into(audio, tx).map_err(|_| Error::Whisper) {
+            producer.send(Err(err)).await;
+            canc_token.cancel();
+        }
+
+        loop {
+            select! {
+                _ = canc_token.cancelled() => {
+                    break
+                }
+                res = rx.recv() => {
+                    match res {
+                        Some(transcribed) => {
+                            let text = transcribed.text().to_owned();
+                            let probability_of_no_speech = transcribed.probability_of_no_speech();
+                            let offset = transcribed.start();
+                            let mut duration = transcribed.duration();
+
+                            //transcribed.duration() seems to be fixed at 30.
+                            //This can brake progress count when the last segment real duration <30
+                            if offset + duration > audio_duration_secs {
+                                duration = audio_duration_secs - offset;
+                            }
+
+                            producer.send(Ok(Event::TranscriptionSegment{text, probability_of_no_speech, offset, duration})).await;
+
+                            if audio_duration_secs >= 0.1 && audio_duration_secs.ceil() > duration.ceil() {
+                                let progress = (duration * 100. / audio_duration_secs).floor() as u8;
+                                producer.send(Ok(Event::DeltaProgress(progress))).await;
+                            }
+                        },
+                        None => {
+                            producer.send(Ok(Event::Progress(100))).await;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    });
+    Ok(())
+}
+
+fn load_audio(path: &Option<PathBuf>) -> Result<Decoder<BufReader<File>>> {
+    match path {
+        Some(path) => {
+            let file = BufReader::new(File::open(path)?);
+            Decoder::new(file).map_err(Into::into)
+        }
+        None => Err(Error::Decoder(DecoderError::UnrecognizedFormat)),
+    }
+}
+
+async fn get_audio_duration_secs(path: &Option<PathBuf>) -> Result<f64> {
+    let mut duration_string = "".to_owned();
+    if let Some(path) = path {
+        let sidecar = Command::new_sidecar("ffprobe")?.args([
+            "-show_entries",
+            "format=duration",
+            "-v",
+            "quiet",
+            "-print_format",
+            "json",
+            path.to_str().unwrap_or(""),
+        ]);
+        let (mut rx, _) = sidecar.spawn().unwrap();
+        while let Some(event) = rx.recv().await {
+            if let CommandEvent::Stdout(line) = event {
+                if line.contains("duration") {
+                    if let Some((_, value)) = line.split_once(':') {
```

Comment:
Instead to parse the output to match the string can we parse the json response, I think we can avoid problems doing like that. Ref. https://github.com/fourviere/fourviere-podcast/blob/main/packages/app/src/native/audio.ts

Reply:
Solved, thx 👍